lib/logstash/inputs/s3.rb

```diff
@@ -91,189 +129,165 @@ def register
-    end
-
-    if @access_key_id.nil? or @secret_access_key.nil?
-      raise ArgumentError.new('Missing AWS credentials')
-    end
-
-    if @bucket.nil?
-      raise ArgumentError.new('Missing AWS bucket')
-    end
-
-    if @sincedb_path.nil?
-      if ENV['HOME'].nil?
-        raise ArgumentError.new('No HOME or sincedb_path set')
-      end
-      @sincedb_path = File.join(ENV["HOME"], ".sincedb_" + Digest::MD5.hexdigest("#{@bucket}+#{@prefix}"))
-    end
-
-    s3 = AWS::S3.new(
-      :access_key_id => @access_key_id,
-      :secret_access_key => @secret_access_key,
-      :region => @region_endpoint
-    )
-
-    @s3bucket = s3.buckets[@bucket]
-
-    unless @backup_to_bucket.nil?
-      @backup_bucket = s3.buckets[@backup_to_bucket]
-      unless @backup_bucket.exists?
-        s3.buckets.create(@backup_to_bucket)
-      end
     end
 
-    unless @backup_to_dir.nil?
-      Dir.mkdir(@backup_to_dir, 0700) unless File.exists?(@backup_to_dir)
+    if @credentials
+      s3 = AWS::S3.new(
+        :access_key_id => @access_key_id,
+        :secret_access_key => @secret_access_key,
+        :region => @region
+      )
+    else
+      s3 = AWS::S3.new(aws_options_hash)
     end
+  end
 
-  end # def register
+  public
+  def aws_service_endpoint(region)
+    return { :s3_endpoint => region }
+  end
 
   public
   def run(queue)
-    loop do
-      process_new(queue)
-      sleep(@interval)
+    Stud.interval(@interval) do
+      process_files(queue)
     end
-    finished
   end # def run
 
   private
-  def process_new(queue, since=nil)
-
-    if since.nil?
-        since = sincedb_read()
-    end
-
-    objects = list_new(since)
-    objects.each do |k|
-      @logger.debug("S3 input processing", :bucket => @bucket, :key => k)
-      lastmod = @s3bucket.objects[k].last_modified
-      process_log(queue, k)
-      sincedb_write(lastmod)
-    end
+  def process_files(queue, since=nil)
+    objects = list_new_files
+    objects.each do |key|
+      @logger.debug("S3 input processing", :bucket => @bucket, :key => key)
 
-  end # def process_new
+      lastmod = @s3bucket.objects[key].last_modified
 
-  private
-  def list_new(since=nil)
+      process_log(queue, key)
 
-    if since.nil?
-      since = Time.new(0)
+      @sincedb.write(lastmod)
     end
+  end # def process_files
 
+  public
+  def list_new_files
     objects = {}
+
     @s3bucket.objects.with_prefix(@prefix).each do |log|
-      if log.last_modified > since
-        objects[log.key] = log.last_modified
+      @logger.debug("S3 input: Found key", :key => log.key)
+
+      unless ignore_filename?(log.key)
+
+        if @sincedb.newer?(log.last_modified)
+          objects[log.key] = log.last_modified
+          @logger.debug("S3 input: Adding to objects[]", :key => log.key)
+        end
       end
     end
-
     return sorted_objects = objects.keys.sort {|a,b| objects[a] <=> objects[b]}
```

Comment:
AWS will return the files (keys?) in alphabetical order and we have to sort them by `last_modification`. This allow us to keep track what was the last file we have processed.